src/symmetric_crypto.cpp

```diff
@@ -0,0 +1,382 @@
+    std::unique_ptr<CipherMemoryStrategyI> _bufferStrategy = nullptr;
+};
+
+
+AESCipher::AESCipher(SymmetricCipherMode mode,
+                     SymmetricCipherPadding padding,
+                     const std::vector<uint8_t> &secretKey,
+                     const std::vector<uint8_t> &iv,
+                     Operation operation)
+{
+    _impl = std::make_unique<AESCipher::Impl>(mode, padding, secretKey, iv, operation);
+}
+
+void AESCipher::update(const std::vector<uint8_t> &message) { _impl->update(message); }
+
+std::vector<uint8_t> AESCipher::read(size_t length) { return _impl->read(length); }
+
+std::vector<uint8_t> AESCipher::readAll() { return _impl->readAll(); }
+
+std::vector<uint8_t> AESCipher::finish()
+{
+    return _impl->finish();
+}
+
+std::vector<uint8_t> AESCipher::getIV() { return _impl->getIV(); }
+
+
+AuthenticatedAESCipher::AuthenticatedAESCipher(SymmetricCipherMode mode,
+                                               SymmetricCipherPadding padding,
+                                               const std::vector<uint8_t> &secretKey,
+                                               const std::vector<uint8_t> &iv,
+                                               const std::vector<uint8_t> &authTag,
+                                               AESCipher::Operation operation)
+        : AESCipher(mode, padding, secretKey, iv, operation)
+{
+    _impl->setAuthTag(authTag);
+}
+
+std::vector<uint8_t> AuthenticatedAESCipher::getAuthTag()
+{
+    return _impl->getAuthTag();
+}
+
+void AuthenticatedAESCipher::setAuthTag(const std::vector<uint8_t> &tag)
+{
+    _impl->setAuthTag(tag);
+}
+
+
+AESCipherBuilder::AESCipherBuilder(SymmetricCipherMode mode,
+                                   std::vector<uint8_t> secretKey)
+        : _mode{mode}, _secretKey{std::move(secretKey)}
+{
+}
+
+AESCipherBuilder &AESCipherBuilder::setIV(std::vector<uint8_t> const &iv)
+{
+    _iv = iv;
+    return *this;
+}
+
+AESCipherBuilder &AESCipherBuilder::setPadding(SymmetricCipherPadding padding)
+{
+    _padding = padding;
+    return *this;
+}
+
+AESCipherBuilder &AESCipherBuilder::setAuthTag(std::vector<uint8_t> const &authTag)
+{
+    _authTag = authTag;
+    return *this;
+}
+
+bool isAuthenticatedCipherMode(SymmetricCipherMode mode)
+{
+    switch (mode) {
+        case SymmetricCipherMode::GCM:
+            return true;
+        case SymmetricCipherMode::CTR:
+            //[[fallthrough]];
+        case SymmetricCipherMode::CBC:
+            return false;
+    }
+}
+
+std::shared_ptr<AESCipher> AESCipherBuilder::buildEncryptor()
+{
+    if (isAuthenticatedCipherMode(_mode)) {
+        throw MoCOCrWException("Specified cipher supports authenticated encryption."
+                               " buildAuthenticatedEncryptor() should be used instead.");
+    }
+    auto cipher = new AESCipher{_mode, _padding, _secretKey, _iv,
+                                AESCipher::Operation::Encryption};
+    return std::shared_ptr<AESCipher>(cipher);
```

Comment:
I'm assuming `std::make_shared<AESCipher>` does not work because the AESCipher c'tor is private and that's why you haven't used it here?

Is this exception-safe, i.e. will the allocated memory be freed if the AESCipher c'tor throws?

Reply:
yes, EVP_CIPHER_CTX is wrapped into a smart pointer and will be freed.